sound/soc/soc-core.c

```diff
@@ -2036,12 +2036,14 @@ static void soc_check_tplg_fes(struct snd_soc_card *card)
 
 static int soc_cleanup_card_resources(struct snd_soc_card *card)
 {
+	/* remove dai links before freeing snd_card */
+	soc_remove_dai_links(card);
+
```

Comment:
how do you deal with pending operations if they are handled in snd_card_free()?

Reply:
@plbossart I'm not sure how to handle this at all. The problem is we need to unload topology before the snd_card is freed. Let me think about it a bit more.

Reply:
@plbossart the only way I can think of avoiding this is by calling snd_soc_tplg_component_remove(component, SND_SOC_TPLG_INDEX_ALL);

in snd_sof_device_remove() for SOF and in skl_remove() for the legacy driver to make sure that the topoogy is unloaded before the card is freed (this would be instead of calling it in the component driver remove()). What do you think about that?

Reply:
@plbossart i've change PR to call remove component in unbind_card instead. Let me know what you think